base/compiler/ssair/passes.jl

```diff
@@ -952,9 +1016,72 @@ function sroa_mutables!(ir::IRCode, defuses::IdDict{Int, Tuple{SPCSet, SSADefUse
         # error at runtime, but is not illegal to have in the IR.
         ismutabletype(typ) || continue
         typ = typ::DataType
+        # First check for any add_finalizer calls
+        add_finalizer_idx = nothing
+        for use in defuse.uses
+            if use.kind === :add_finalizer
+                # For now: Only allow one add_finalizer per allocation
+                add_finalizer_idx !== nothing && @goto skip
+                add_finalizer_idx = use.idx
+            end
+        end
+        if add_finalizer_idx !== nothing
+            # For now: Require that all uses and defs are in the same basic block,
+            # so that live range calculations are easy.
+            bb = ir.cfg.blocks[block_for_inst(ir.cfg, first(defuse.uses).idx)]
+            minval = typemax(Int)
+            maxval = 0
```

Comment:
This makes the code type-unstable (due to closure conversion). It seems rare to reach here though, so is that okay?

Reply:
Will add type annotations just in case.